### PR TITLE
s3select/tests: replace with queries

### PR DESCRIPTION
**Pull request by albin-antony**

### CODE REVIEW

File: test/s3select_test_json_parser.cpp [hunk 1]
```diff
@@ -317,38 +274,491 @@ std::string json2 = R"({
...
+{ "type": "home5", "number": "734928_5" },
+{ "type": "home6", "number": "734928_6" },
+{ "type": "home7", "number": "734928_7" },
+{ "type": "home8", "number": "734928_8" },
+{ "type": "home9", "number": "734928_9" }
+]
+}
+)";
 
 #define TEST5 \
```
Comment (galsalomon66) [inline test/s3select_test_json_parser.cpp:763]: please remove TEST5 / base64

File: test/s3select_test_json_parser.cpp [hunk 2]
```diff
@@ -362,137 +772,6 @@ c3RhcnJheSI6IHsKICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg
...
-
-#define TEST11 \
-"ewoiZmlyc3ROYW1lIjogIkpvZSIsCiJsYXN0TmFtZSI6ICJKYWNrc29uIiwKImdlbmRlciI6ICJtYWxlIiwKImFnZSI6ICJ0d2VudHkiLAoiYWRkcmVzcyI6IHsKInN0cmVldEFkZHJlc3MiO\
-iAiMTAxIiwKImNpdHkiOiAiU2FuIERpZWdvIiwKInN0YXRlIjogIkNBIgp9LAoicGhvbmVOdW1iZXJzIjogWwp7ICJ0eXBlIjogImhvbWUxIiwgIm51bWJlciI6ICI3MzQ5MjgyXzEiIH0sCn\
-sgInR5cGUiOiAiaG9tZTIiLCAibnVtYmVyIjogIjczNDkyODJfMiIgfSwKeyAidHlwZSI6ICJob21lMyIsICJudW1iZXIiOiAiNzM0OTI4XzMiIH0sCnsgInR5cGUiOiAiaG9tZTQiLCAibnV\
-tYmVyIjogIjczNDkyOF80IiB9LAp7ICJ0eXBlIjogImhvbWU1IiwgIm51bWJlciI6ICI3MzQ5MjhfNSIgfSwKeyAidHlwZSI6ICJob21lNiIsICJudW1iZXIiOiAiNzM0OTI4XzYiIH0sCnsg\
-InR5cGUiOiAiaG9tZTciLCAibnVtYmVyIjogIjczNDkyOF83IiB9LAp7ICJ0eXBlIjogImhvbWU4IiwgIm51bWJlciI6ICI3MzQ5MjhfOCIgfSwKeyAidHlwZSI6ICJob21lOSIsICJudW1iZ\
-XIiOiAiNzM0OTI4XzkiIH0KXQp9Cg=="
-
 std::string run_sax(const char * in)
```
Comment (galsalomon66) [inline test/s3select_test_json_parser.cpp:775]: no need for ` switch(key_value._type()) ` 
can use `to_string`

File: test/s3select_test_json_parser.cpp [hunk 3]
```diff
@@ -656,12 +935,30 @@ int sax_row_count(const char *in, std::vector<std::string>& from_clause)
...
+  }
+
+  s3selectEngine::json_object json_query_processor(&s3select_syntax);
+  status = json_query_processor.run_s3select_on_stream(result, json_input.data(), json_input.size(), json_input.size());
+  status = json_query_processor.run_s3select_on_stream(result, 0, 0, json_input.size());
+
+  return status;
+}
+
 TEST(TestS3selectJsonParser, sax_vs_dom)
```
Comment (galsalomon66) [inline test/s3select_test_json_parser.cpp:956]: please remove not-in-use TEST's

**Comment by galsalomon66** [comment]
please address comments.
good coverage, it mainly covers row-definition, i.e. the row-count per from-clause.
the array-size are the same for most arrays, maybe it's better to change that, to make sure it counts the correct array.

**Comment by galsalomon66** [comment]
this PR can be closed.
no need for a rebase.
need to merged only `test/s3select_test_json_parser.cpp`

File: test/s3select_test.cpp [hunk 4]
```diff
@@ -2827,3 +2872,30 @@ phoneNumbers.addr. : 88
...
+TEST(TestS3selectFunctions, csv_json_count)
+{
+  std::string input;
+  std::string result;
+  size_t size = 5;
+  generate_csv(input, size);
+  std::string js = convert_to_json(input.c_str(), input.size());
+  const char* input_query_1 = "select count(0) from s3object[*];";
+
+  run_json_query(input_query_1, js, result);
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2884]: the concept of converting CSV to JSON is about running 
the same query on the same input (the JSON input is a conversion from CSV)
and compared both results.
in that way, we gain coverage for all SQL operators.

as mentioned before the SQL statement also needs a conversion.

File: test/s3select_test_json_parser.cpp (hunk 1 above)
Comment (galsalomon66) [reply test/s3select_test_json_parser.cpp:766]: is it removed ?

File: test/s3select_test.cpp [hunk 5]
```diff
@@ -2827,3 +2915,32 @@ phoneNumbers.addr. : 88
...
+  std::string input;
+  std::string result;
+  size_t size = 5;
+  generate_csv(input, size);
+  std::string js = convert_to_json(input.c_str(), input.size());
+  std::string input_query_1 = "select count(0) from s3object;";
+  const char* json_query_1 = convert_query(input_query_1);
+
+  run_json_query(json_query_1, js, result);
+  ASSERT_EQ(result, "5");
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2929]: it needs to compare that CSV result equal to JSON result just as in Parquet (Parquet = JSON)

File: include/s3select.h [hunk 6]
```diff
@@ -2699,7 +2699,7 @@ class json_object : public base_s3object
       JsonHandler.process_json_buffer(0, 0, true);//TODO end-of-stream = end-of-row
       m_end_of_stream = true;
       sql_execution_on_row_cb();
-      result = s3select_result;
+      result += s3select_result;
```
Comment (galsalomon66) [inline include/s3select.h:2702]: this is not correct 
it cause results duplication

File: test/s3select_test.h [hunk 7]
```diff
@@ -284,6 +284,88 @@ int run_query_on_parquet_file(const char* input_query, const char* input_file, s
...
+
+const char* convert_query(std::string& expression)
+{
+  std::string from_clause = "s3object";
+  boost::replace_all(expression, from_clause, "s3object[*].root");
+
+  std::string from_clause_1 = "stdin";
+  boost::replace_all(expression, from_clause_1, "s3object[*].root");
+
+  std::string col_1 = "_1";
```
Comment (galsalomon66) [inline test/s3select_test.h:348]: these lines repeat themselves.
its better to reduce that.

File: test/s3select_test.h [hunk 8]
```diff
@@ -525,18 +612,39 @@ std::string run_s3select_opserialization_quot(std::string expression,std::string
...
+  int status = s3select_syntax.parse_query(json_query);
+  if (status != 0)
+  {
+    std::cout << "failed to parse query " << s3select_syntax.get_error_description() << std::endl;
+    return -1;
+  }
+
+  json_object json_query_processor(&s3select_syntax);
+  status = json_query_processor.run_s3select_on_stream(result, json_input.data(), json_input.size(), json_input.size());
+  status = json_query_processor.run_s3select_on_stream(result, 0, 0, json_input.size());
```
Comment (galsalomon66) [inline test/s3select_test.h:629]: the second call to `run_s3select_on_stream` override the previous results

File: test/s3select_test_json_parser.cpp [hunk 9]
```diff
@@ -656,13 +952,24 @@ int sax_row_count(const char *in, std::vector<std::string>& from_clause)
...
+  int status = s3select_syntax.parse_query(json_query);
+  if (status != 0)
+  {
+    std::cout << "failed to parse query " << s3select_syntax.get_error_description() << std::endl;
+    return -1;
+  }
 
+  s3selectEngine::json_object json_query_processor(&s3select_syntax);
+  status = json_query_processor.run_s3select_on_stream(result, json_input.data(), json_input.size(), json_input.size());
+  status = json_query_processor.run_s3select_on_stream(result, 0, 0, json_input.size());
```
Comment (galsalomon66) [inline test/s3select_test_json_parser.cpp:968]: the second call to `run_s3select_on_stream` might overrides the previous result.

File: test/s3select_test.cpp [hunk 10]
```diff
@@ -1241,25 +1250,26 @@ TEST(TestS3selectFunctions, in_expressions)
 
   ASSERT_NE(s3select_result_9,failure_sign);
 
-  const std::string input_query_10 = "select int(_1) from s3object where _1 like \"_3\";";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1244]: in this type of query, one option is to manually create the converted-JSON-query and pass it to `run_s3select`

File: test/s3select_test.cpp [hunk 11]
```diff
@@ -1692,15 +1702,15 @@ TEST(TestS3selectFunctions, predicate_as_projection_column)
 
   ASSERT_EQ(count_2,0);
 
-  const std::string input_query_3 = "select (_1 like \"_3\") from s3object where character_length(_1) = 2 and substring(_1,2,1) in (\"3\");";
+  //const std::string input_query_3 = "select (_1 like \"_3\") from s3object where character_length(_1) = 2 and substring(_1,2,1) in (\"3\");";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1705]: same.
manually converting the JSON query.

File: test/s3select_test.cpp [hunk 12]
```diff
@@ -1916,11 +1926,11 @@ TEST(TestS3selectFunctions, truefalse_in_expressions)
 
   ASSERT_NE(s3select_result_9,failure_sign);
 
-  const std::string input_query_10 = "select int(_1) from s3object where (_1 like \"_3\") = true;";
+  //const std::string input_query_10 = "select int(_1) from s3object where (_1 like \"_3\") = true;";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:1929]: same.
manually converting the JSON query.

File: test/s3select_test.cpp [hunk 13]
```diff
@@ -2649,11 +2659,11 @@ TEST(TestS3selectFunctions, presto_syntax_alignments)
 
   auto s3select_res = run_s3select(input_query, input);
 
-  const std::string input_presto_query = "Select t._1,t._2 fRom s3OBJECT t whEre _1 = _2";
+  //const std::string input_presto_query = "Select t._1,t._2 fRom s3OBJECT t whEre _1 = _2";
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2662]: this query verifies some syntax definitions which are critical for integration with presto (could be also other utilities).

in this case, it's possible to skip the JSON verification.
some modifications to the syntax parser are required.

File: test/s3select_test.cpp [hunk 14]
```diff
@@ -2850,3 +2860,36 @@ phoneNumbers.addr. : 88
 
 }
 
+TEST(TestS3selectFunctions, csv_json_count)
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2863]: why this TEST is required?

File: test/s3select_test.cpp [hunk 15]
```diff
@@ -2850,3 +2860,36 @@ phoneNumbers.addr. : 88
...
+  std::string csv_query_1 = "select count(0) from s3object;";
+  std::string s3select_result_1 = run_s3select(csv_query_1, input);
+
+  const char* json_query_1 = convert_query(csv_query_1);
+  run_json_query(json_query_1, js, result);
+
+  ASSERT_EQ(result, s3select_result_1);
+}
+
+TEST(TestS3selectFunctions, sum_json)
```
Comment (galsalomon66) [inline test/s3select_test.cpp:2879]: same.

File: test/s3select_test.h [hunk 16]
```diff
@@ -284,6 +284,88 @@ int run_query_on_parquet_file(const char* input_query, const char* input_file, s
...
+            ss << "\"c" << ++counter << "\"" << ":";
+            ss << "\"";
+            ss << *m_stream;
+            previous = false;
+          } else if (*m_stream != column_delimiter) {
+              ss << *m_stream;
+          } else if (*m_stream == column_delimiter) {
+              if (previous)  {
+                ss << "\"c" << ++counter << "\"" << ":";
+                //ss << "\"";
```
Comment (galsalomon66) [inline test/s3select_test.h:315]: please remove comment

**Comment by galsalomon66** [comment]
the related-presto test is commented out.

File: test/s3select_test_json_parser.cpp (hunk 9 above)
Comment (galsalomon66) [reply test/s3select_test_json_parser.cpp:968]: what about this?

**Comment by albin-antony** [comment]
> the related-presto test is commented out.

@galsalomon66 Addressed it. Please take a look

**Pull request merged**